plugins/modules/identity/keycloak/keycloak_user.py

```diff
@@ -0,0 +1,527 @@
+                         options=dict(
+                             # CredentialRepresentation
+                             priority=dict(type='int'),
+                             type=dict(type='str'),
+                             temporary=dict(type='bool'),
+                             value=dict(type='str'),
+                         )),
+        disableable_credential_types=dict(type='list', elements='str', aliases=[
+                                          'disableableCredentialTypes']),
+        email=dict(type='str'),
+        email_verified=dict(type='bool', aliases=['emailVerified']),
+        enabled=dict(type='bool'),
+        first_name=dict(type='str', aliases=['firstName']),
+        groups=dict(type='list', elements='str'),
+        id=dict(type='str'),
+        last_name=dict(type='str', aliases=['lastName']),
+        not_before=dict(type='int', aliases=['notBefore']),
+        origin=dict(type='str'),
+        realm_roles=dict(type='list', elements='str', aliases=['realmRoles']),
+        required_actions=dict(type='list', elements='str',
+                              aliases=['requiredActions']),
+        service_account_client_id=dict(
+            type='str', aliases=['serviceAccountClientId']),
+    )
+
+    argument_spec.update(meta_args)
+
+    module = AnsibleModule(argument_spec=argument_spec,
+                           supports_check_mode=True,
+                           required_one_of=([['id', 'username'],
+                                             ['token', 'auth_realm', 'auth_username', 'auth_password']]),
+                           required_together=([['auth_realm', 'auth_username', 'auth_password']]))
+
+    result = dict(changed=False, msg='', diff={}, user='')
+
+    # Obtain access token, initialize API
+    try:
+        connection_header = get_token(module.params)
+    except KeycloakError as e:
+        module.fail_json(msg=str(e))
+
+    kc = KeycloakAPI(module, connection_header)
+
+    realm = module.params.get('realm')
+    state = module.params.get('state')
+    username = module.params.get('username')
+    attributes = module.params.get('attributes')
+
+    # attributes in Keycloak have their values returned as lists
+    # via the API. attributes is a dict, so we'll transparently convert
+    # the values to lists.
+    if attributes is not None:
+        for key, val in module.params['attributes'].items():
+            module.params['attributes'][key] = [
+                val] if not isinstance(val, list) else val
+
+    # Filter and map the parameters names that apply to the user
+    user_params = [x for x in module.params
+                   if x not in list(keycloak_argument_spec().keys()) + ['state', 'realm'] and module.params.get(x) is not None]
+
+    found_users = kc.get_users_by_username(username, realm=realm)
+
+    if found_users is None or found_users == []:
+        before_user = {}
+    else:
+        before_user = found_users[0]
```

Comment:
The documents above indicate say that `username` is only required when doing a create, but I can't (from the code) see how the value of `before_user` (which is used below when deleting a user) is populated properly when the username is not supplied.